proto/docs/p4runtime_type_serialization.md

```diff
@@ -0,0 +1,96 @@
+and to define a custom serialization format for complex P4_16 types. The
+serialization would maybe be trivial for header types but would require some
+work for header unions, stacks, etc. Rather than coming-up with a serialization
+format from scratch, we decided to use a Protobuf representation for all P4_16
+types, since the P4Runtime service already uses Protobuf messages.
+
+## P4 type specifications in p4info.proto
+In order for the P4Runtime client to generate correctly-formatted messages and
+for the P4Runtime service implementation to validate them, P4Info needs to
+specify the type of each P4 expression which is exposed to the control-plane. In
+the register example above, client and server need to know that each element of
+the register has type `ip_t`, which is a header union with 2 possible headers:
+`ipv4` with type `ipv4_t` and `ipv6` with type `ipv6`. Similarly, they need to
+know the field layout of both these header types.
+
+To achieve this we introduce 2 main protobuf messages: `P4TypeInfo` and
+`P4DataTypeSpec`.
+
+`P4TypeInfo` is a top-level member of `P4Info` and includes Protobuf maps
+storing the type specification for all the named types in the P4_16
+program. These named types are `struct`, `header`, `header_union` and `enum`;
+for each of these we have a type specification message, respectively
+`P4StructTypeSpec`, `P4HeaderTypeSpec`, `P4HeaderUnionTypeSpec` and
+`P4EnumTypeSpec`. We support P4 annotations for named types, which is useful to
+identify well-known headers, such as IPv4 or IPv6. `P4TypeInfo` also include the
+list of parser errors for the program, as a `P4ErrorTypeSpec` message.
+
+`P4DataTypeSpec` is meant to be used in `P4Info`, everywhere where the P4Runtime
+client can provide a value for a P4_16 expression. `P4DataTypeSpec` describes
+the compile-time of the expression as a Protobuf `oneof`, which is a `string` in
+case of a named type (`struct`, `header`, `header_union` or `enum`), an empty
+Protobuf message for `bool` and `error`, and a Protobuf message for other
+anonymous types (`bit<W>`, `int<W>`, `varbit<W>`, `tuple` or `stack`). The
+"bitstring" types (`bit<W>`, `int<W>` and `varbit<W>`) are grouped together in
+the `P4BitstringLikeTypeSpec` message, since they are the only sub-types allowed
+in headers and values with one of these types are represented similarly in
+P4Runtime (with binary strings).
+
+## P4 data in p4runtime.proto
+P4Runtime uses the `P4Data` message to represent values with arbitrary
+types. The P4Runtime client must generate correct `P4Data` messages based on the
+type specification information included in `P4Info`. The `P4Data` message was
+designed to introduce little overhead compared to using binary strings in the
+most common case (P4_16 `bit<W>` type).
+
+Just like its `P4Info` counterpart `P4DataTypeSpec`, `P4Data` uses a `oneof` to
+represent all possible values.
+
+### `enum` and `error`
+We currently use human-readable `string` in `P4Data` to represent `enum` and
```

Comment:
People want the ability to specify this in the P4 program - perhaps using an annotation.
It would make sense to allow this - sometimes wasting 32 bits on enums may be too expensive.